pulser/sampler/sampler.py

```diff
@@ -0,0 +1,279 @@
+
+    return d
+
+
+def _prepare_dict(seq: Sequence, N: int) -> dict:
+    """Constructs empty dict of size N.
+
+    Usually N is the duration of seq.
+    """
+
+    def new_qty_dict() -> dict:
+        return {
+            "amp": np.zeros(N),
+            "det": np.zeros(N),
+            "phase": np.zeros(N),
+        }
+
+    def new_qdict() -> dict:
+        return {qubit: new_qty_dict() for qubit in seq._qids}
+
+    if seq._in_xy:
+        return {
+            "Global": {"XY": new_qty_dict()},
+            "Local": {"XY": new_qdict()},
+        }
+    else:
+        return {
+            "Global": defaultdict(new_qty_dict),
+            "Local": defaultdict(new_qdict),
+        }
+
+
+def _write_dict(
+    seq: Sequence,
+    samples: dict[str, list[QubitSamples]],
+    addrs: dict[str, str],
+) -> dict:
+    """Export the given samples to a nested dictionary."""
+    # Get the duration
+    if not _same_duration(samples):
+        raise ValueError("All the samples do not share the same duration.")
+    N = list(samples.values())[0][0].amp.size
+
+    d = _prepare_dict(seq, N)
+
+    for ch_name, some_samples in samples.items():
+        basis = seq.declared_channels[ch_name].basis
+        addr = addrs[ch_name]
+        if addr == "Global":
+            # Take samples on only one qubit and write them
+            a_qubit = next(iter(seq._qids))
+            to_write = [x for x in some_samples if x.qubit == a_qubit]
+            for s in to_write:
+                d["Global"][basis]["amp"] += s.amp
+                d["Global"][basis]["det"] += s.det
+                d["Global"][basis]["phase"] += s.phase
+        else:
+            for s in some_samples:
+                d["Local"][basis][s.qubit]["amp"] += s.amp
+                d["Local"][basis][s.qubit]["det"] += s.det
+                d["Local"][basis][s.qubit]["phase"] += s.phase
+    return d
+
+
+def _same_duration(samples: dict[str, list[QubitSamples]]) -> bool:
+    durations: list[int] = []
+    flatten_samples: list[QubitSamples] = []
+    for some_samples in samples.values():
+        flatten_samples.extend(some_samples)
+    for s in flatten_samples:
+        durations.extend((s.amp.size, s.det.size, s.phase.size))
+    return durations.count(durations[0]) == len(durations)
+
+
+def _sample_channel(
+    seq: Sequence, ch_name: str, strategy: TimeSlotExtractionStrategy
+) -> list[QubitSamples]:
+    """Compute a list of QubitSamples for a channel."""
+    qs: list[QubitSamples] = []
+    grouped_slots = strategy(seq._schedule[ch_name])
+
+    for group in grouped_slots:
+        ss = _sample_slots(seq.get_duration(), *group)
+        qs.extend(ss)
+    return qs
+
+
+def _sample_slots(N: int, *slots: _TimeSlot) -> list[QubitSamples]:
+    """Gather samples of a list of _TimeSlot in a single Samples instance."""
+    # Same target in one group, guaranteed by the strategy (this seems
+    # weird, it's not enforced by the structure,bad design?)
```

Comment:
`_sample_slots` expects that `*slots` is a list of `_TimeSlot` that all have the same targets. I hence extract them from only the first element with

```python
qubits = slots[0].targets
```

So I find the overall pattern I have written a bit risky because

1. Nothing prevent to call `_sample_slots()` with slots having different target
2. The correct way to gather `_TimeSlots` is with a `TimeSlotExtractionStrategy` before feeding them to `_sample_slots()`, but this type alias does say nor enforce anything about grouping by same targets. So one may implement a strategy satisfying the prototype, but not respecting the grouping, handing bad samples down the line.

I think it's no big deal for this PR, but it's just a reminder that the design is not satisfactory.

Reply:
Ok, I see. Thanks for the clarification!